Add post on groovy testing (#3017)
oktadeveloper/okta.github.io: _source/_posts/2019-08-21-make-java-tests-groovy-again.md

---
layout: blog_post
title: Make Java Tests Groovy Again With Hamcrest

> description: 'A tutorial on testing Java with Groovy'

Reviewer: ```suggestion
description: 'A tutorial on testing Java code with Groovy.'
```

> tags: spring, oauth, oauth 2.0, java, spring boot, testing, hamcrest, groovy

Reviewer: This should be an array
```suggestion
tags: ]spring, oauth, oauth 2.0, java, spring boot, testing, hamcrest, groovy]
```

             key2: "value2",
             nested: [
                 anotherKey: "anotherValue"
             ]]

def myList = ["one", "two", "three"]
```

The Elvis operator makes ternary expressions shorter:

```groovy
def displayName = user.name ?: 'Anonymous'
```

String interpolation! Seriously, why do we not have this not in Java yet?

```groovy
def answer = 42
println "The answer to the ultimate question of life, " +
        "the universe and everything is ${answer}"
```
Results in:
```txt
The answer to the ultimate question of life, the universe and everything is 42
```

Groovy has support for default method parameter values, and while you can accomplish the same thing in Java, it is more verbose. For example, in Groovy we could write:

```groovy
String sayHello(String name = "Joe", String greeting = "How are you?") {
    println "Hello ${name}, ${greeting}"
}
```

The Java equivalent would be:

```java
public String sayHello() {
    return sayHello("Joe");
}

public String sayHello(String name) {
    return sayHello(name, "How are you?");
}

public String sayHello(String name, String greeting) {
    System.out.println("Hello " + name + ", " greeting);
}
```

> The Groovy block is faster to write more concise.

Reviewer: and more concise?

Ever seen anything like:
```java
/**
* This method is exposed for testing only, it may change in the future without notice.
**/
@VisibleForTesting
public Object iWishThisWasPrivate() {
    // ...
}
```
I think it's better to avoid this scenario entirely when possible.

## Add OAuth 2.0 Support to the Application

Now that we have tested our application, we can go ahead and start it. If you don't already have a free Okta Developer account, signup at [developer.okta.com](https://developer.okta.com/signup)

Once you have an account, take note of your Okta domain.  This will be in your initial activation email, as well as in the top right of your Okta Developer Console after you log in.

To create an OAuth/OIDC application, from your Okta Developer Console, click **Applications** menu, then click the green **Add Application** button.

Select **Web** as the platform and click **Next**.

Name your application, something like "Testing is Groovy"

Change the Login redirect URIs to `http://localhost:8080/login/oauth2/code/okta`.

Click **Done**.

Take note of the Client ID and Client Secret on the next page and update the  `src/main/resources/application.yml`.

```yaml
okta:
  oauth2:

>     issuer: {yourOktaDomain}/oauth2/default

Reviewer: ```suggestion
    issuer: https://{yourOktaDomain}/oauth2/default
```

Author: ugg, I cannot believe I missed that one, i couldn't wait until we switched all the refs to `https://{yourOktaDomain}`
